src/qpx_scan_algo.cpp

```diff
@@ -0,0 +1,940 @@
+/*
+ * This file is part of the QPxTool project.
+ * Copyright (C) 2007-2009, Gennady "ShultZ" Kozlov <qpxtool@mail.ru>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ * See the file "COPYING" for the exact licensing terms.
+ *
+ */
+
+#include <cstdio>
+#include <cstdlib>
+#include <errno.h>
+
+#include "qpx_scan.hpp"
+#include "threads.hpp"
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "qpx_writer.hpp"
+
+#define EMAXLINE 1024
+
+#define USE_FFLUSH
+
+// void qscanner::show_avg(struct timeval s, struct timeval e, long lba)
+void qscanner::show_avg_speed(long lba) {
+  double btime;
+  int spdKB;
+  float spdX;
+  btime = (e.tv_sec - s.tv_sec) + (e.tv_usec - s.tv_usec) / 1000000.0;
+  spdKB = (int)(((lba - lba_sta + 1) << 1) / btime);
+  spdX = spdKB / (float)spd1X;
+  printf("\nTest time: %6.2fs\navg speed: %5.3f X  %5d kB/s\n", btime, spdX,
+         spdKB);
+}
+
+void qscanner::calc_cur_speed(long sects) {
+  double btime;
+  //	int spdKB;
+  //	float spdX;
+  btime =
+      (blke.tv_sec - blks.tv_sec) + (blke.tv_usec - blks.tv_usec) / 1000000.0;
+  spdKB = (int)((sects << 1) / btime);
+  spdX = spdKB / (float)spd1X;
+  //	printf("\nTest time: %6.2fs\navg speed: %5.3f X  %5d kB/s\n", btime,
+  // spdX, spdKB);
+}
+
+int qscanner::readline(int fd, char *buf, int maxlen) {
+  int cnt = 0;
+  char *cbuf = buf;
+  int r;
+  int sret;
+  fd_set rd_set;
+  timeval tv;
+
+  FD_ZERO(&rd_set);
+  if (tchar >= 0) {
+    cbuf[0] = tchar;
+    cnt++;
+    cbuf++;
+    tchar = -1;
+  }
+  while (!stop_req && (cnt < (maxlen - 1))) {
+    FD_SET(fd, &rd_set);
+    tv.tv_sec = 1;
+    tv.tv_usec = 0;
+
+#ifndef _WIN32
+    sret = select(fd + 1, &rd_set, NULL, NULL, &tv);
+#else
+    //		sret = WaitForSingleObject((void*)fd, 1000);
+    //		errno = GetLastError();
+    sret = 1;
+    errno = 0;
+#endif
+    //		printf("readline select(): %d, errno = [%d] %s\n", sret, errno,
+    // strerror(errno));
+    if (sret < 0) {
+      printf("select():  %s\n", strerror(errno));
+      if (errno == EINTR)
+        continue;
+      return -1;
+    }
+
+    else if (sret > 0 && FD_ISSET(fd, &rd_set)) {
```

Comment:
Codacy found an issue: [Condition 'sret>0' is always true](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)